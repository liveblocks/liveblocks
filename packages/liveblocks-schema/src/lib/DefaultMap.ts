/**
 * Like ES6 map, but takes a default factory function which will be used
 * to create entries for missing keys on the fly.
 *
 * Useful for code like:
 *
 *   const map = new DefaultMap(() => []);
 *   map.get('foo').push('hello');
 *   map.get('foo').push('world');
 *   map.get('foo')
 *   // ['hello', 'world']
 *
 */
export default class DefaultMap<K, V> extends Map<K, V> {
  #_factoryFn?: (key: K) => V;

  constructor(
    factoryFn?: (key: K) => V,
    entries?: readonly (readonly [K, V])[] | null
  ) {
    super(entries);
    this.#_factoryFn = factoryFn;
  }

  get #factoryFn() {
    if (this.#_factoryFn === undefined) {
      throw new Error("DefaultMap used without a factory function");
    }
    return this.#_factoryFn;
  }

  /**
   * Gets the value at the given key, or creates it.
   *
   * Difference from normal Map: if the key does not exist, it will be created
   * on the fly using the factory function, and that value will get returned
   * instead of `undefined`.
   */
  getOrCreate(key: K, factoryFn?: (key: K) => V): V {
    let value = super.get(key);
    if (value === undefined) {
      value = factoryFn !== undefined ? factoryFn(key) : this.#factoryFn(key);
      this.set(key, value);
    }
    return value;
  }

  /**
   * Sets the value at the given key.
   *
   * Difference from normal Map: if the second arg is a function, it will be
   * called with the current value as its argument and the return value will be
   * set instead. If the returned value is `undefined`, it will remove the key,
   * meaning the next time the key is accessed it will get re-created by the
   * factory function.
   */
  set(key: K, value: V): this;
  set(key: K, valueFn: (prev: V) => V | undefined): this;
  set(key: K, value: V | ((prev: V) => V | undefined)): this {
    if (typeof value === "function") {
      value = (value as (prev: V) => V)(this.get(key) ?? this.#factoryFn(key));
    }

    if (value === undefined) {
      super.delete(key);
      return this;
    } else {
      return super.set(key, value);
    }
  }
}
