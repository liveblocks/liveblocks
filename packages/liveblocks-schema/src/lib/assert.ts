export function assertNever(_value: never, msg: string): never {
  throw new Error(msg);
}
