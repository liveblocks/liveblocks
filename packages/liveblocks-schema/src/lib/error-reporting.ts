import type { Range } from "../ast";
import { prettify } from "../prettify";
import { indent } from "./indent";

const WHITESPACE_ONLY_RE = /^\s*$/;

// TODO: Use an ANSI color library when available
// This is still useful to do when in node on an OS terminal, but not for error
// messages sent over HTTP or within a browser env
const nocolor = (value: string): string => value;
const colors = {
  cyan: nocolor,
  gray: nocolor,
  red: nocolor,
  yellow: nocolor,
};

export class Source {
  path: string;
  contents: string;

  static fromText(text: string): Source {
    return new Source("<string>", text);
  }

  private constructor(path: string, contents: string) {
    this.path = path;
    this.contents = contents;
  }
}

export type LineInfo = {
  offset: number;
  line1: number; // 1-based line number
  column1: number; // 1-based column number
};

/**
 * For a string like "foo\nbarbaz\nquxxx\n", builds: [4, 11, 17]
 */
function buildOffsetLUT(lines: string[]): number[] {
  const offsets = [];
  let total = 0;
  for (const line of lines) {
    total += line.length + 1;
    offsets.push(total);
  }
  return offsets;
}

/**
 * Helper to report inline error messages in source files.
 *
 * Given (at least a) filename, will be able to print error messages in a human
 * friendly way, annotating the source file.
 *
 *     ErrorReporter.fromPath('/path/to/file.lsl')
 *     ErrorReporter.fromText('x = 42')
 *
 */
export class ErrorReporter {
  src: Source;
  #lines: string[] | undefined;
  #offsets: number[] | undefined;
  #hasErrors: boolean = false;

  // static fromPath(path: string): ErrorReporter {
  //   return new ErrorReporter(Source.fromPath(path));
  // }

  static fromText(programText: string): ErrorReporter {
    return new ErrorReporter(Source.fromText(programText));
  }

  static fromSrc(src: Source): ErrorReporter {
    return new ErrorReporter(src);
  }

  private constructor(src: string | Source) {
    this.src = typeof src === "string" ? Source.fromText(src) : src;
  }

  get hasErrors(): boolean {
    return this.#hasErrors;
  }

  contents(): string {
    return this.src.contents;
  }

  lines(): string[] {
    if (!this.#lines) {
      this.#lines = this.contents().split("\n");
    }
    return this.#lines;
  }

  offsets(): number[] {
    if (!this.#offsets) {
      this.#offsets = buildOffsetLUT(this.lines());
    }
    return this.#offsets;
  }

  lineInfo(offset: number): LineInfo {
    const lines = this.lines();
    const offsets = this.offsets();
    const lineno0 = offsets.findIndex((eol) => offset < eol);
    if (lineno0 < 0) {
      // Offset out of bounds, point to the very last character
      return { offset, line1: lines.length, column1: 1 };
    } else if (lineno0 === 0) {
      return { offset, line1: 1, column1: offset + 1 };
    } else {
      const preveol = offsets[lineno0 - 1];
      const column0 = offset - preveol;
      return { offset, line1: lineno0 + 1, column1: column0 + 1 };
    }
  }

  *iterAnnotateSourceLines(
    startOffset: number,
    endOffset: number
  ): Generator<string> {
    const lines = this.lines();
    const start = this.lineInfo(startOffset);
    const end = this.lineInfo(endOffset);

    // We want to show a bit of leading and trailing context from the
    // source file. To that end, we'll start at most 3 lines before the
    // line with the error on it, and try to "walk back" line by line until
    // we hit the line with the error info, or find a non-empty line. This
    // will effectively strip off any "leading empty lines" from the
    // context.
    let startLineno1 = Math.max(1, start.line1 - 3);
    while (startLineno1 < start.line1) {
      const line = lines[startLineno1 - 1];
      if (WHITESPACE_ONLY_RE.test(line)) {
        startLineno1++;
      } else {
        break;
      }
    }

    // Ditto, but for trailing empty lines
    let endLineno1 = Math.min(end.line1 + 3, lines.length);
    while (endLineno1 > end.line1) {
      const line = lines[endLineno1 - 1];
      if (WHITESPACE_ONLY_RE.test(line)) {
        endLineno1--;
      } else {
        break;
      }
    }

    for (let lineno = startLineno1; lineno <= endLineno1; lineno++) {
      // Print the line itself
      const line = lines[lineno - 1];
      yield `${colors.gray(lineno.toString().padStart(5, " "))}  ${line}`;

      // Print an annotation if this is for the right line
      if (lineno === start.line1) {
        if (start.line1 === end.line1) {
          yield indent(
            6 + start.column1,
            colors.red("^".repeat(end.column1 - start.column1))
          );
        } else {
          yield indent(6 + start.column1, colors.red("^"));
        }
      }
    }
  }

  formatHeading(left: string, right: string): string {
    const twidth: number | null =
      typeof process !== "undefined" ? process.stdout.columns ?? null : null;

    // If we have access to the terminal's width, use it to spread out the
    // left and right part, otherwise just stick 'em together
    if (twidth) {
      const outerPadding = 2;
      const innerPadding = 1;
      const padding = outerPadding + innerPadding;
      const spacer = "â€•".repeat(
        twidth - (left.length + padding) - (right.length + padding)
      );
      return `  ${left} ${spacer} ${right}`;
    } else {
      return `  ${left} ${right}`;
    }
  }

  *iterLongParseError(message: string, range?: Range): Generator<string> {
    this.#hasErrors = true;
    const [startOffset, endOffset] = range ?? [undefined, undefined];

    yield "";
    yield colors.cyan(this.formatHeading("Parse error", `in ${this.src.path}`));
    if (startOffset !== undefined && endOffset !== undefined) {
      yield* this.iterAnnotateSourceLines(startOffset, endOffset);
    }
    yield "";
    yield indent(2, colors.red(message));
    yield "";
  }

  *iterShortError(message: string, range?: Range): Generator<string> {
    this.#hasErrors = true;
    const [startOffset] = range ?? [undefined, undefined];

    if (startOffset !== undefined) {
      // Strip the trailing period from the message, if any
      if (message.endsWith(".")) {
        message = message.slice(0, -1);
      }

      const start = this.lineInfo(startOffset);
      yield `${message} (at ${start.line1}:${start.column1})`;
    } else {
      yield message;
    }
  }

  getShortParseError(message: string, range?: Range): string {
    return Array.from(this.iterShortError(message, range)).join("\n");
  }

  throwParseError(message: string, range?: Range): never {
    const err = new Error(this.getShortParseError(message, range));
    err.name = "ParseError";
    throw err;
  }

  printParseError(message: string, range?: Range): void {
    for (const line of this.iterLongParseError(message, range)) {
      console.log(line);
    }
  }

  *iterLongSemanticError(
    title: string,
    description: (string | null)[],
    range?: Range
  ): Generator<string> {
    this.#hasErrors = true;
    const [startOffset, endOffset] = range ?? [undefined, undefined];

    yield "";
    yield colors.cyan(
      this.formatHeading("Schema error", ` in ${this.src.path}`)
    );
    if (title) {
      yield "";
      yield indent(2, title);
    }

    if (startOffset !== undefined && endOffset !== undefined) {
      yield "";
      yield* this.iterAnnotateSourceLines(startOffset, endOffset);
    }

    yield "";
    for (const desc of description) {
      if (desc === null) {
        continue;
      }
      const text =
        typeof desc === "string"
          ? desc
          : colors.yellow("    " + prettify(desc));
      yield indent(2, text);
    }
    yield "";
  }

  getShortSemanticError(title: string, range?: Range): string {
    return Array.from(this.iterShortError(title, range)).join("\n");
  }

  throwSemanticError(title: string, range?: Range): never {
    const err = new Error(this.getShortSemanticError(title, range));
    err.name = "SemanticError";
    throw err;
  }

  printSemanticError(
    title: string,
    description: (string | null)[],
    range?: Range
  ): void {
    for (const line of this.iterLongSemanticError(title, description, range)) {
      console.log(line);
    }
  }
}
