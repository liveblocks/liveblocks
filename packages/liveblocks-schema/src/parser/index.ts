import type { Document, Node, TypeExpr } from "../ast";
import type { Source } from "../lib/error-reporting";
import { ErrorReporter } from "../lib/error-reporting";
import * as generatedParser from "./generated-parser";

export type StartRule =
  // See the PEG grammar. These are the valid start rule to kick off the
  // parsing process.  All other grammar rules are internal to the parser.
  "Document" | "TypeExpr";

function parseGrammarRule(
  src: string | Source | ErrorReporter,
  startRule: StartRule
): Node | Node[] {
  const reporter =
    typeof src === "string"
      ? ErrorReporter.fromText(src)
      : "path" in src
      ? ErrorReporter.fromSrc(src)
      : src;

  try {
    return generatedParser.parse(reporter.contents(), { startRule });
  } catch (err_: unknown) {
    const e = err_ as Error;

    /**
     * If this is a parse error (due to a syntax error), report this in
     * a visually pleasing manner in the console.
     */
    if (/SyntaxError/.test(e.name)) {
      type SyntaxError = Error & {
        location: {
          start: { offset: number; line: number; column: number };
          end: { offset: number; line: number; column: number };
        };
      };
      const se = e as SyntaxError;
      reporter.throwParseError(se.message, [
        se.location.start.offset,
        se.location.end.offset,
      ]);
    } else {
      throw e;
    }
  }

  throw new Error("Should never get here");
}

export function parseDocument(src: string | Source | ErrorReporter): Document {
  return parseGrammarRule(src, "Document") as Document;
}

export function parseTypeExpr(src: string | Source | ErrorReporter): TypeExpr {
  return parseGrammarRule(src, "TypeExpr") as TypeExpr;
}
