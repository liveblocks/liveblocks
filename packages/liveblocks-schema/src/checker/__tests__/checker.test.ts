import type * as AST from "../../ast";
import { ErrorReporter } from "../../lib/error-reporting";
import { parseDocument } from "../../parser";
import { check } from "..";

describe("checker", () => {
  it("type checking a full document", () => {
    const schemaText = `
      type IamStatic {}

      type Storage {
        x: IamStatic
        y: LiveObject<IamLive>
      }

      type IamLive {}
    `;

    const reporter = ErrorReporter.fromText(schemaText);
    const output = check(parseDocument(reporter), reporter);

    const range = expect.any(Array);

    expect(output).toEqual({
      root: {
        _kind: "ObjectTypeDefinition",
        name: {
          _kind: "TypeName",
          name: "Storage",
          range,
        },
        isStatic: false,
        fields: [
          {
            _kind: "FieldDef",
            name: {
              _kind: "Identifier",
              name: "x",
              range,
            },
            optional: false,
            type: {
              _kind: "TypeRef",
              asLiveObject: false,
              range,
              ref: {
                _kind: "TypeName",
                name: "IamStatic",
                range,
              },
            },
            range,
          },
          {
            _kind: "FieldDef",
            name: {
              _kind: "Identifier",
              name: "y",
              range,
            },
            optional: false,
            type: {
              _kind: "TypeRef",
              ref: {
                _kind: "TypeName",
                name: "IamLive",
                range,
              },
              asLiveObject: true,
              range,
            },
            range,
          },
        ],
        range,
      },
      definitions: expect.any(Array),
      getDefinition: expect.anything(),
    });

    const x = output.root.fields[0].type as AST.TypeRef;
    expect(output.getDefinition(x)).toEqual({
      _kind: "ObjectTypeDefinition",
      name: {
        _kind: "TypeName",
        name: "IamStatic",
        range,
      },
      isStatic: true,
      fields: [],
      range,
    });

    const y = output.root.fields[1].type as AST.TypeRef;
    expect(output.getDefinition(y)).toEqual({
      _kind: "ObjectTypeDefinition",
      name: {
        _kind: "TypeName",
        name: "IamLive",
        range,
      },
      isStatic: false,
      fields: [],
      range,
    });
  });

  it("large document (snapshot test)", () => {
    const schemaText = `
      // union Shape = Rect | Circle

      type RGB { r: Int, g: Int, b: Int }

      type Rect {
        type: String  # TODO: Use "rect" here
        x: Int
        y: Int
        width: Int
        height: Int
        fill?: RGB
        stroke?: RGB
      }

      type Circle {
        type: String // TODO: Use "circle" here
        cx: Int
        cy: Int
        radius: Int
        fill ? : RGB
        stroke?: RGB
      }

      type Storage {
        // shapes: LiveList<Shape>
        mycircle: LiveObject<Circle>
        myrect: LiveObject<Rect>
      }
    `;

    const reporter = ErrorReporter.fromText(schemaText);
    const output = check(parseDocument(reporter), reporter);
    expect(output).toMatchSnapshot();
  });
});
