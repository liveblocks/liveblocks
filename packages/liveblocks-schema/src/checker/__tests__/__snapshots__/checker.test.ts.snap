// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checker large document (snapshot test) 1`] = `
{
  "ast": {
    "_kind": "Document",
    "definitions": [
      {
        "_kind": "ObjectTypeDefinition",
        "fields": [
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "r",
              "range": [
                56,
                57,
              ],
            },
            "optional": false,
            "range": [
              56,
              62,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "IntType",
              "range": [
                59,
                62,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "g",
              "range": [
                64,
                65,
              ],
            },
            "optional": false,
            "range": [
              64,
              70,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "IntType",
              "range": [
                67,
                70,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "b",
              "range": [
                72,
                73,
              ],
            },
            "optional": false,
            "range": [
              72,
              78,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "IntType",
              "range": [
                75,
                78,
              ],
            },
          },
        ],
        "isStatic": true,
        "leadingComment": null,
        "name": {
          "_kind": "TypeName",
          "name": "RGB",
          "range": [
            50,
            53,
          ],
        },
        "range": [
          45,
          80,
        ],
      },
      {
        "_kind": "ObjectTypeDefinition",
        "fields": [
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "type",
              "range": [
                108,
                112,
              ],
            },
            "optional": false,
            "range": [
              108,
              145,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "StringType",
              "range": [
                114,
                145,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "x",
              "range": [
                154,
                155,
              ],
            },
            "optional": false,
            "range": [
              154,
              160,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "IntType",
              "range": [
                157,
                160,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "y",
              "range": [
                169,
                170,
              ],
            },
            "optional": false,
            "range": [
              169,
              175,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "IntType",
              "range": [
                172,
                175,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "width",
              "range": [
                184,
                189,
              ],
            },
            "optional": false,
            "range": [
              184,
              194,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "IntType",
              "range": [
                191,
                194,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "height",
              "range": [
                203,
                209,
              ],
            },
            "optional": false,
            "range": [
              203,
              214,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "IntType",
              "range": [
                211,
                214,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "fills",
              "range": [
                223,
                228,
              ],
            },
            "optional": true,
            "range": [
              223,
              236,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "ArrayExpr",
              "ofType": {
                "_kind": "TypeRef",
                "asLiveObject": false,
                "range": [
                  231,
                  234,
                ],
                "ref": {
                  "_kind": "TypeName",
                  "name": "RGB",
                  "range": [
                    231,
                    234,
                  ],
                },
              },
              "range": [
                231,
                236,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "strokes",
              "range": [
                245,
                252,
              ],
            },
            "optional": true,
            "range": [
              245,
              268,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "LiveListExpr",
              "ofType": {
                "_kind": "TypeRef",
                "asLiveObject": false,
                "range": [
                  264,
                  267,
                ],
                "ref": {
                  "_kind": "TypeName",
                  "name": "RGB",
                  "range": [
                    264,
                    267,
                  ],
                },
              },
              "range": [
                255,
                268,
              ],
            },
          },
        ],
        "isStatic": false,
        "leadingComment": null,
        "name": {
          "_kind": "TypeName",
          "name": "Rect",
          "range": [
            93,
            97,
          ],
        },
        "range": [
          88,
          276,
        ],
      },
      {
        "_kind": "ObjectTypeDefinition",
        "fields": [
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "type",
              "range": [
                306,
                310,
              ],
            },
            "optional": false,
            "range": [
              306,
              345,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "StringType",
              "range": [
                312,
                345,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "cx",
              "range": [
                354,
                356,
              ],
            },
            "optional": false,
            "range": [
              354,
              361,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "IntType",
              "range": [
                358,
                361,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "cy",
              "range": [
                370,
                372,
              ],
            },
            "optional": false,
            "range": [
              370,
              377,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "IntType",
              "range": [
                374,
                377,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "radius",
              "range": [
                386,
                392,
              ],
            },
            "optional": false,
            "range": [
              386,
              397,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "IntType",
              "range": [
                394,
                397,
              ],
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "fill",
              "range": [
                406,
                410,
              ],
            },
            "optional": true,
            "range": [
              406,
              418,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "TypeRef",
              "asLiveObject": false,
              "range": [
                415,
                418,
              ],
              "ref": {
                "_kind": "TypeName",
                "name": "RGB",
                "range": [
                  415,
                  418,
                ],
              },
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "stroke",
              "range": [
                427,
                433,
              ],
            },
            "optional": true,
            "range": [
              427,
              439,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "TypeRef",
              "asLiveObject": false,
              "range": [
                436,
                439,
              ],
              "ref": {
                "_kind": "TypeName",
                "name": "RGB",
                "range": [
                  436,
                  439,
                ],
              },
            },
          },
        ],
        "isStatic": false,
        "leadingComment": null,
        "name": {
          "_kind": "TypeName",
          "name": "Circle",
          "range": [
            289,
            295,
          ],
        },
        "range": [
          284,
          447,
        ],
      },
      {
        "_kind": "ObjectTypeDefinition",
        "fields": [
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "mycircle",
              "range": [
                513,
                521,
              ],
            },
            "optional": false,
            "range": [
              513,
              541,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "TypeRef",
              "asLiveObject": true,
              "range": [
                523,
                541,
              ],
              "ref": {
                "_kind": "TypeName",
                "name": "Circle",
                "range": [
                  534,
                  540,
                ],
              },
            },
          },
          {
            "_kind": "FieldDef",
            "leadingComment": null,
            "name": {
              "_kind": "Identifier",
              "name": "myrect",
              "range": [
                550,
                556,
              ],
            },
            "optional": false,
            "range": [
              550,
              574,
            ],
            "trailingComment": null,
            "type": {
              "_kind": "TypeRef",
              "asLiveObject": true,
              "range": [
                558,
                574,
              ],
              "ref": {
                "_kind": "TypeName",
                "name": "Rect",
                "range": [
                  569,
                  573,
                ],
              },
            },
          },
        ],
        "isStatic": false,
        "leadingComment": null,
        "name": {
          "_kind": "TypeName",
          "name": "Storage",
          "range": [
            460,
            467,
          ],
        },
        "range": [
          455,
          582,
        ],
      },
    ],
    "range": [
      7,
      582,
    ],
  },
  "definitions": [
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "r",
            "range": [
              56,
              57,
            ],
          },
          "optional": false,
          "range": [
            56,
            62,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "IntType",
            "range": [
              59,
              62,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "g",
            "range": [
              64,
              65,
            ],
          },
          "optional": false,
          "range": [
            64,
            70,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "IntType",
            "range": [
              67,
              70,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "b",
            "range": [
              72,
              73,
            ],
          },
          "optional": false,
          "range": [
            72,
            78,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "IntType",
            "range": [
              75,
              78,
            ],
          },
        },
      ],
      "isStatic": true,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "RGB",
        "range": [
          50,
          53,
        ],
      },
      "range": [
        45,
        80,
      ],
    },
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "type",
            "range": [
              108,
              112,
            ],
          },
          "optional": false,
          "range": [
            108,
            145,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "StringType",
            "range": [
              114,
              145,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "x",
            "range": [
              154,
              155,
            ],
          },
          "optional": false,
          "range": [
            154,
            160,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "IntType",
            "range": [
              157,
              160,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "y",
            "range": [
              169,
              170,
            ],
          },
          "optional": false,
          "range": [
            169,
            175,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "IntType",
            "range": [
              172,
              175,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "width",
            "range": [
              184,
              189,
            ],
          },
          "optional": false,
          "range": [
            184,
            194,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "IntType",
            "range": [
              191,
              194,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "height",
            "range": [
              203,
              209,
            ],
          },
          "optional": false,
          "range": [
            203,
            214,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "IntType",
            "range": [
              211,
              214,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "fills",
            "range": [
              223,
              228,
            ],
          },
          "optional": true,
          "range": [
            223,
            236,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "ArrayExpr",
            "ofType": {
              "_kind": "TypeRef",
              "asLiveObject": false,
              "range": [
                231,
                234,
              ],
              "ref": {
                "_kind": "TypeName",
                "name": "RGB",
                "range": [
                  231,
                  234,
                ],
              },
            },
            "range": [
              231,
              236,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "strokes",
            "range": [
              245,
              252,
            ],
          },
          "optional": true,
          "range": [
            245,
            268,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "LiveListExpr",
            "ofType": {
              "_kind": "TypeRef",
              "asLiveObject": false,
              "range": [
                264,
                267,
              ],
              "ref": {
                "_kind": "TypeName",
                "name": "RGB",
                "range": [
                  264,
                  267,
                ],
              },
            },
            "range": [
              255,
              268,
            ],
          },
        },
      ],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "Rect",
        "range": [
          93,
          97,
        ],
      },
      "range": [
        88,
        276,
      ],
    },
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "type",
            "range": [
              306,
              310,
            ],
          },
          "optional": false,
          "range": [
            306,
            345,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "StringType",
            "range": [
              312,
              345,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "cx",
            "range": [
              354,
              356,
            ],
          },
          "optional": false,
          "range": [
            354,
            361,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "IntType",
            "range": [
              358,
              361,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "cy",
            "range": [
              370,
              372,
            ],
          },
          "optional": false,
          "range": [
            370,
            377,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "IntType",
            "range": [
              374,
              377,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "radius",
            "range": [
              386,
              392,
            ],
          },
          "optional": false,
          "range": [
            386,
            397,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "IntType",
            "range": [
              394,
              397,
            ],
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "fill",
            "range": [
              406,
              410,
            ],
          },
          "optional": true,
          "range": [
            406,
            418,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              415,
              418,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "RGB",
              "range": [
                415,
                418,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "stroke",
            "range": [
              427,
              433,
            ],
          },
          "optional": true,
          "range": [
            427,
            439,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": false,
            "range": [
              436,
              439,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "RGB",
              "range": [
                436,
                439,
              ],
            },
          },
        },
      ],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "Circle",
        "range": [
          289,
          295,
        ],
      },
      "range": [
        284,
        447,
      ],
    },
    {
      "_kind": "ObjectTypeDefinition",
      "fields": [
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "mycircle",
            "range": [
              513,
              521,
            ],
          },
          "optional": false,
          "range": [
            513,
            541,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": true,
            "range": [
              523,
              541,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Circle",
              "range": [
                534,
                540,
              ],
            },
          },
        },
        {
          "_kind": "FieldDef",
          "leadingComment": null,
          "name": {
            "_kind": "Identifier",
            "name": "myrect",
            "range": [
              550,
              556,
            ],
          },
          "optional": false,
          "range": [
            550,
            574,
          ],
          "trailingComment": null,
          "type": {
            "_kind": "TypeRef",
            "asLiveObject": true,
            "range": [
              558,
              574,
            ],
            "ref": {
              "_kind": "TypeName",
              "name": "Rect",
              "range": [
                569,
                573,
              ],
            },
          },
        },
      ],
      "isStatic": false,
      "leadingComment": null,
      "name": {
        "_kind": "TypeName",
        "name": "Storage",
        "range": [
          460,
          467,
        ],
      },
      "range": [
        455,
        582,
      ],
    },
  ],
  "getDefinition": [Function],
  "root": {
    "_kind": "ObjectTypeDefinition",
    "fields": [
      {
        "_kind": "FieldDef",
        "leadingComment": null,
        "name": {
          "_kind": "Identifier",
          "name": "mycircle",
          "range": [
            513,
            521,
          ],
        },
        "optional": false,
        "range": [
          513,
          541,
        ],
        "trailingComment": null,
        "type": {
          "_kind": "TypeRef",
          "asLiveObject": true,
          "range": [
            523,
            541,
          ],
          "ref": {
            "_kind": "TypeName",
            "name": "Circle",
            "range": [
              534,
              540,
            ],
          },
        },
      },
      {
        "_kind": "FieldDef",
        "leadingComment": null,
        "name": {
          "_kind": "Identifier",
          "name": "myrect",
          "range": [
            550,
            556,
          ],
        },
        "optional": false,
        "range": [
          550,
          574,
        ],
        "trailingComment": null,
        "type": {
          "_kind": "TypeRef",
          "asLiveObject": true,
          "range": [
            558,
            574,
          ],
          "ref": {
            "_kind": "TypeName",
            "name": "Rect",
            "range": [
              569,
              573,
            ],
          },
        },
      },
    ],
    "isStatic": false,
    "leadingComment": null,
    "name": {
      "_kind": "TypeName",
      "name": "Storage",
      "range": [
        460,
        467,
      ],
    },
    "range": [
      455,
      582,
    ],
  },
}
`;
